File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    public int[][] directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    private int rows;
    private int cols;
    public int numIslands(char[][] grid) {
        int numislands = 0;
        rows = grid.length;
        cols = grid[0].length;
        Set<String> visited = new HashSet<>();
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                String pos = r + ", " + c;
                if (grid[r][c] == '1' && !visited.contains(pos)) {
                    numislands++;
                    bfs(grid, r, c, visited);
                }
            }
        }
        return numislands;

    }

    private void bfs(char[][] grid, int r, int c, Set<String> visited) {
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{r, c});

        while (!q.isEmpty()) {
            int[] coord = q.poll();
            int row = coord[0];
            int col = coord[1];
            for (int[] direction : directions) {
                int newr = row + direction[0];
                int newc = col + direction[1];
                if (newr >= 0 && newr < rows && newc >= 0 && newc < cols && grid[newr][newc] == '1' && !visited.contains(newr + ", " + newc)){
                    q.add(new int[]{newr, newc});
                    visited.add(newr + ", " + newc);
                }
            }
        }
        
    }
}

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        numIslands = 0
        rows, cols = len(grid), len(grid[0])
        visited = set()
        def bfs(r: int, c: int):
            q = deque()
            visited.add((r, c))
            q.append((r, c))
            directions = [(0, 1), (1, 0), (-1, 0), (0, -1)]
            while q:
                row, col = q.popleft()
                for direction in directions:
                    newr = row + direction[0]
                    newc = col + direction[1]
                    if newr >= 0 and newr < rows and newc >= 0 and newc < cols and grid[newr][newc] == '1' and not (newr, newc) in visited:
                        visited.add((newr, newc))
                        q.append((newr, newc))
                        

        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == '1' and not (r, c) in visited:
                    numIslands += 1
                    bfs(r, c)

       
        
        return numIslands
                

        
